examples/step-95/step-95.cc

```diff
@@ -0,0 +1,1707 @@
+
+    ConditionalOStream pcout;
+
+    // MatrixFree object
+    MatrixFree<dim, double> matrix_free;
+
+    PoissonOperator<dim> poisson_operator;
+
+    // We need two separate DoFHandlers. The first manages the DoFs for the
+    // discrete level set function that describes the geometry of the domain.
+    std::unique_ptr<FE_Q<dim>> fe_level_set;
+    DoFHandler<dim>            level_set_dof_handler;
+    VectorType                 level_set;
+
+    // The second DoFHandler manages the DoFs for the solution of the Poisson
+    // equation.
+    hp::FECollection<dim> fe_collection;
+    DoFHandler<dim>       dof_handler;
+    VectorType            solution;
+
+    NonMatching::MeshClassifier<dim> mesh_classifier;
+
+    VectorType rhs;
+
+    const MappingQ<dim> mapping;
+
+    // NonMatching::MappingInfo objects
+    std::unique_ptr<NonMatching::MappingInfo<dim, dim, VectorizedArrayType>>
+      mapping_info_cell;
+    std::unique_ptr<NonMatching::MappingInfo<dim, dim, VectorizedArrayType>>
+      mapping_info_surface;
+    std::unique_ptr<NonMatching::MappingInfo<dim, dim, VectorizedArrayType>>
+      mapping_info_faces;
+
+    const unsigned int dof_index  = 0;
+    const unsigned int quad_index = 0;
+
+    bool is_dg;
+  };
+
+
+
+  template <int dim>
+  PoissonSolver<dim>::PoissonSolver()
+    : fe_degree(1)
+    , rhs_function(4.0)
+    , triangulation(MPI_COMM_WORLD)
+    , pcout(std::cout,
+            Utilities::MPI::this_mpi_process(
+              triangulation.get_communicator()) == 0)
+    , level_set_dof_handler(triangulation)
+    , dof_handler(triangulation)
+    , mesh_classifier(level_set_dof_handler, level_set)
+    , mapping(1)
+    , is_dg(false)
+  {}
+
+
+
+  // @sect3{Setting up the Background Mesh}
+  // We generate a background mesh with perfectly Cartesian cells. Our domain is
+  // a unit disc centered at the origin, so we need to make the background mesh
+  // a bit larger than $[-1, 1]^{\text{dim}}$ to completely cover $\Omega$.
+  template <int dim>
+  void PoissonSolver<dim>::make_grid()
+  {
+    pcout << "Creating background mesh" << std::endl;
+
+    triangulation.clear();
+    GridGenerator::hyper_cube(triangulation, -1.21, 1.21);
+    triangulation.refine_global(2);
+  }
+
+
+
+  // @sect3{Setting up the Discrete Level Set Function}
+  // The discrete level set function is defined on the whole background mesh.
+  // Thus, to set up the DoFHandler for the level set function, we distribute
+  // DoFs over all elements in $\mathcal{T}_h$. We then set up the discrete
+  // level set function by interpolating onto this finite element space.
+  template <int dim>
+  void PoissonSolver<dim>::setup_discrete_level_set()
+  {
+    pcout << "Setting up discrete level set function" << std::endl;
+
+    fe_level_set = std::make_unique<FE_Q<dim>>(fe_degree);
+    level_set_dof_handler.distribute_dofs(*fe_level_set);
+    level_set.reinit(level_set_dof_handler.n_dofs());
```

Comment:
@bergbauer In this tutorial, you set up the DoF layout for the level-set vector in a "matrix-based" way, even though this is a matrix-free tutorial. I understand this is likely necessary because the MeshClassifier loops over all non-artificial cells (see https://github.com/dealii/dealii/blob/master/source/non_matching/mesh_classifier.cc#L365-L384) and retrieves all level-set DoF values for classification (noting that the matrix-free container only allocates some of the locally relevant DoFs for parallel vectors).

Would it make sense to add a comment explaining why the matrix-free object isn’t used to initialize the DoF vector? We ran into this issue ourselves when trying to use the level-set vector from a matrix-free level-set solver in a multi-physics cut problem. It might help clarify this for others as well.

Reply:
> PS: I just noticed that you don't use a parallel layout here, but I think my question still makes sense :-)

Thanks for noticing, @mschreter! I forgot about the level set when parallelizing the tutorial.

I switched to a parallel layout and added a comment. It would also be possible to setup the level set matrix-free with `MatrixFree::AdditionalData::hold_all_faces_to_owned_cells` enabled, but since we don't advect the level set here, let's stick with the classical `reinit` function.